### PR TITLE
Only emit one copy-to-host or copy-to-accelerator compiler warning per source location, even if it has multiple host uses.

### PR DESCRIPTION
**Pull request by mhong**
Resolves [SR-8412](https://bugs.swift.org/browse/SR-8412).

### CODE REVIEW

**Comment by mhong** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/TFPartition.cpp [hunk 1]
```diff
@@ -606,6 +606,31 @@ void BlocksReachingTensorCode::dump() {
...
 //===----------------------------------------------------------------------===//
 
+namespace llvm {
+template<typename T> struct DenseMapInfo;
+
+// An alternative is to SourceManager::getLineAndColumn() as the
+// DenseMap/DenseSet key type instead of SourceRange. That would require that
+// the call-site translate a SILLocation / SourceRange to line and column
+// numbers first.
+template<> struct DenseMapInfo<SourceRange> {
```
Comment (lattner) [inline lib/SILOptimizer/Mandatory/TFPartition.cpp]: This should go in the header that defines SourceRange, and upstreamed :-)

File: lib/SILOptimizer/Mandatory/TFPartition.cpp [hunk 2]
```diff
@@ -942,7 +972,10 @@ diagnoseCopyToAccelerator(SILValue value, SILInstruction *user,
 
   auto &ctx = hostFn.getModule().getASTContext();
 
-  // Emit the warning.
+  // Emit the warning on this value, if that has not been done before.
+  if (!copyToAccelWarningLocs.insert(loc.getSourceRange()).second)
+    return;
+
   diagnose(ctx, loc.getSourceLoc(), diagID, description)
```
Comment (lattner) [inline lib/SILOptimizer/Mandatory/TFPartition.cpp]: Do you need to unique based on the entire source range?  Would it be fine to just unique based on loc.getSourceLoc() ?   That would eliminate the need for template specialization entirely.

**Review by lattner — APPROVED** [review]

**Review by rxwei — APPROVED** [review]
Nice fix! LGTM.

File: lib/SILOptimizer/Mandatory/TFPartition.cpp [hunk 3]
```diff
@@ -606,6 +606,31 @@ void BlocksReachingTensorCode::dump() {
...
 
+namespace llvm {
+template<typename T> struct DenseMapInfo;
+
+// An alternative is to SourceManager::getLineAndColumn() as the
+// DenseMap/DenseSet key type instead of SourceRange. That would require that
+// the call-site translate a SILLocation / SourceRange to line and column
+// numbers first.
+template<> struct DenseMapInfo<SourceRange> {
+  static SourceRange getEmptyKey() { return SourceRange(); }
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFPartition.cpp]: Consider making the empty key and the tombstone key different?

File: lib/SILOptimizer/Mandatory/TFPartition.cpp (hunk 2 above)
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFPartition.cpp]: `DenseMapInfo` is not defined for `SourceLoc` either, so I might as make `SourceRange` work as the key type.

Compiler error for using `llvm::DenseSet<SourceLoc> copyToHostWarningLocs`:
```
/usr/local/google/home/hongm/ssd_part/git/swift-base/llvm/include/llvm/ADT/DenseMap.h:421:22: error: no member named 'getEmptyKey' in 'llvm::DenseMapInfo<swift::SourceLoc>'
    return KeyInfoT::getEmptyKey();
```

File: lib/SILOptimizer/Mandatory/TFPartition.cpp (hunk 1 above)
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFPartition.cpp]: Will do!

File: lib/SILOptimizer/Mandatory/TFPartition.cpp (hunk 3 above)
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFPartition.cpp]: Done

**Pull request merged**